[BE-61] feat: 개인정보 등록 기능 구현

xxbeann opened the pull request:
## #️⃣연관된 이슈

> #149 

## 📝작업 내용

> 개인정보(이름, 몸무게, 키, 혈액형)를 등록하는 API 입니다.
현재 정수값만 입력받을 수 있습니다. 소수점으로 입력해도 정수로 저장됩니다.

### 스크린샷 (선택)
<img width="1400" height="553" alt="image" src="https://github.com/user-attachments/assets/e76d4676-7d1b-415d-9a8b-e5ae15bea3ac" />


## 💬리뷰 요구사항(선택)

> PhysicalInfo에서 몸무게와 키를 다시 한 번 검증하는 로직이 추가되는 게 나을까요?


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- New Features
  - 개인 정보 등록 기능 추가: 이름, 체중, 키, 혈액형을 받아 프로필에 신체 정보 저장/수정 가능
  - 신규 엔드포인트 POST /profile/personal-information 제공, 성공 시 표준 응답 반환
- Bug Fixes
- Validation
  - 이름(2–10자 한글), 체중/키(양수, 최대 999), 혈액형(A/B/O/AB) 유효성 검증 강화
  - 잘못된 혈액형 입력 시 명확한 오류 응답 및 메시지 제공
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

xxbeann committed:
feat: PhysicalInfo VO 작성

xxbeann committed:
feat: 개인정보 등록 컨트롤러 작성

xxbeann committed:
feat: 개인정보 요청 DTO 및 검증 메시지 작성

xxbeann committed:
feat: 개인정보 등록 로직 작성

xxbeann committed:
feat: String BloodType을 BloodType으로 변경

xxbeann committed:
feat: 파라미터 타입 변경

xxbeann committed:
chore: 주석제거

xxbeann committed:
feat:updateName에 이름 검증 메서드 추가

Woojae-Jeong commented on src/main/java/com/econo_4factorial/newproject/user/domain/User.java:
```diff
@@ -51,7 +55,13 @@ public User(String email, String name, String appleSub) {
+        userInfo.updateBasicInformation(email,phoneNumber);
+    }
+
+    public void registerPersonalInformation(String name, Long weight, Long height, BloodType bloodType) {
+        if (physicalInfo == null) physicalInfo = new PhysicalInfo();
+        userInfo.updateName(name);
+        physicalInfo.updatePersonalInformation(weight, height, bloodType);
```
JPA의 리플렉션으로 인해 null값이 들어가있어 update 시 NPE 발생이 걱정되신다면, name, weight, height, bloodType을 가진 객체를 그대로 생성해서 대입해주는 방식도 있을 것 같네요. 이는 참고사항입니다!

xxbeann replied:
그런 방법도 있군요!
새 객체를 생성하고 값을 업데이트 하는 방식(제가 작성)
null 여부 상관없이 새 객체를 생성해서 교체하는 게 더 코드 가독성도 좋고 NPE 걱정도 없네요~
반영하겠습니다!

Woojae-Jeong commented on src/main/java/com/econo_4factorial/newproject/common/exception/NullRequestException.java:
```diff
@@ -0,0 +1,7 @@
+    public NullRequestException() {
+        super(CommonErrorType.NULL_REQUEST_EXCEPTION);
+    }
+}
```
`@NotBlank` 어노테이션이 null값 체크도 해주다보니 해당 예외는 불필요 할 것 같습니다

xxbeann replied:
아 맞다 그러네요 깜빡했습니다! 반영할게요!

Woojae-Jeong commented on src/main/java/com/econo_4factorial/newproject/user/domain/BloodType.java:
```diff
@@ -0,0 +1,20 @@
+    O,
+    AB;
+
+    public static BloodType fromString(String bloodType) {
+        if (bloodType == null) throw new NullRequestException();
+        try {
+            return BloodType.valueOf(bloodType.toUpperCase());
+        } catch (IllegalArgumentException e) {
+            throw new InvalidBloodTypeException();
+        }
+    }
```
`fromString()` 메소드를 서비스 내에서 사용할 때 null값이 들어와서 던지는 예외가 커스텀 예외랑 자바 예외 둘 중 어떤 것으로 하는 게 좋은지 고민해 주시면 좋을 것 같아요

xxbeann replied:
reqDTO에 있는 NotBlank어노테이션이 null 체크를 해주더라도 한번 더 검증해주는 게 나을 것 같다는 판단에 서비스에도 `NullRequestException` 예외처리를 했습니다. 실제로 기본정보(전화번호, 이메일)등록 기능 구현중에, req에는 valid 검사를 진행했는데, 영속화하는 단계에서 꼬였는지 값이 반대로 들어가더라구요. 그래서 전화번호에 이메일이 들어가고, 이메일에 이름이 들어가고 이랬던 경험이 있어서 계속 확인을 하는 게 중요하다고 생각했습니다.
그래서 `NullRequestException`으로 예외처리를 한거구요. 제 생각에는 객체가 Null인걸 감지해 예외처리 하는 건 나중에도 쓰일 것 같아서 Common에러로 정의하긴 했습니다...
커스텀 예외와 자바 예외 중 어느게 더 나아 보이시나요?

Woojae-Jeong replied:
일단 이 부분은 개발자마다 의견이 다를 것이고, 저는 자바 예외가 올바르다고 생각하는 사람입니다.
 
프론트로부터 잘못된 요청이 들어오게 된다면 유효성 검사를 통해서 거를 수 있습니다. 다만 해당 fromString()이 꼭 지금 작성하신 개인정보 등록 API에 의해서만 작동하는 건 아닙니다. 다시 말하자면, `동료나 본인이 해당 enum 객체를 어디 코드에서 생성할 줄 모를 뿐더러, 그 과정에서 null값을 넣을 수 있습니다`. 이런 상황에서 400에러를 클라이언트에게 보내주는 것이 올바르지 않다는 것은 이해하실 겁니다. 

그럼 `500에러를 반환하면 되지 않는가?` 라고 생각하실 수 있습니다. 에러가 발생했다는 것은 그 예외를 해석하는 데에 개발자의 노력이 들어갑니다. 상대적으로 친숙한 `IllegalArgumentException`은 개발자가 `어디선가 잘못된 인자값을 줬다`고 빠르게 이해할 수 있지만, 커스텀 예외는 `이 예외가 어떠한 상황에서 발생하도록 설계되어 있는지 알아내야 하는 노력`이 들게 됩니다. 또한 자바가 지원하는 예외인데 별도로 정의한다는 것은 그만큼 예외가 많아진다는 것을 의미합니다. 그렇기에 저는 우리만의 비즈니스 규칙에 의해서 발생되는 예외가 아닌 null 검사, 음수 검사, 양수 검사와 같은 비정상적인 데이터에 대한 검사로 인해 발생하는 예외는 별도로 커스텀 예외를 정의하는 것보다 많이들 알고있는 `IllegalArgumentException` 이 더 가독성이 좋다고 생각합니다. 

이는 참고하면 좋을 것 같아요. 
https://tecoble.techcourse.co.kr/post/2020-08-17-custom-exception/

xxbeann replied:
자세한 답변 감사합니다! 리팩토링 해볼게요~~

Woojae-Jeong reviewed (commented):
리뷰 몇 개 남겼습니다!

xxbeann committed:
refactor: physicalInfo가 null일 경우 빈 객체 생성 후 updatePersonalInformation 호출

- 기존: physicalInfo가 null일 경우 빈 객체 생성 후 updatePersonalInformation 호출
- 변경: 항상 새로운 PhysicalInfo(height, weight, bloodType)를 생성하여 교체
- 장점:
  - NPE 가능성을 근본적으로 제거
  - 코드 단순화 및 가독성 향상
  - VO 특성에 맞는 불변성 패턴 유지

xxbeann committed:
chore: 불필요 import 제거

xxbeann committed:
feat: Null Common Error 제거 및 자바 에러로 리팩토링

Woojae-Jeong commented:
확인했습니다! 머지하도록 할게요!
